### PR TITLE
inline Sigma functions

### PR DESCRIPTION
**Pull request by dcousens**
Before (SHA256)
```
40, 3162, 6513.72, 0.4854368932038835
41, 3868, 6769, 0.5714285714285714
42, 4732, 6719.44, 0.704225352112676
43, 5788, 6771.96, 0.8547008547008547
44, 7079, 6795.84, 1.0416666666666667
45, 8660, 6841.4, 1.2658227848101267
46, 10593, 6885.45, 1.5384615384615385
47, 12957, 6867.21, 1.8867924528301887
48, 15849, 6815.07, 2.3255813953488373
49, 19387, 6785.45, 2.857142857142857
50, 23714, 6742.21568627451, 3.5172413793103448
51, 29007, 6892.752475247525, 4.208333333333333
52, 35481, 6889.514563106796, 5.15
53, 43401, 6875.40594059406, 6.3125
54, 53088, 6833.108910891089, 7.769230769230769
55, 64938, 6803.028571428571, 9.545454545454545
56, 79433, 6808.542857142857, 11.666666666666666
57, 97163, 6801.41, 14.285714285714286
58, 118850, 6856.7307692307695, 17.333333333333332
59, 145378, 6608.090909090909, 22
60, 177828, 6905.941747572816, 25.75
61, 217520, 6851.023622047244, 31.75
62, 266073, 6881.198275862069, 38.666666666666664
63, 325462, 6875.957746478874, 47.333333333333336
64, 398107, 6805.247863247863, 58.5
65, 486968, 6858.704225352113, 71
66, 595662, 6618.466666666666, 90
67, 728618, 6873.754716981132, 106
68, 891251, 6855.776923076923, 130
69, 1090184, 6899.898734177215, 158
70, 1333521, 6873.819587628866, 194
71, 1631173, 6740.384297520661, 242
72, 1995262, 6786.60544217687, 294
73, 2440619, 6874.98309859155, 355
74, 2985383, 6831.54004576659, 437
75, 3651741, 6890.077358490566, 530
76, 4466836, 6850.9754601227, 652
77, 5463865, 6872.786163522012, 795
78, 6683439, 6883.0473738414, 971
79, 8175230, 6801.356073211315, 1202
80, 10000000, 6872.852233676976, 1455
```

After
```
40, 3162, 17232.9, 0.1834862385321101
41, 3868, 17328.64, 0.22321428571428573
42, 4732, 17319.12, 0.273224043715847
43, 5788, 17421.88, 0.33222591362126247
44, 7079, 17555.92, 0.4032258064516129
45, 8660, 17493.2, 0.49504950495049505
46, 10593, 17584.38, 0.6024096385542169
47, 12957, 17103.24, 0.7575757575757576
48, 15849, 17592.39, 0.9009009009009009
49, 19387, 17642.17, 1.098901098901099
50, 23714, 17785.5, 1.3333333333333333
51, 29007, 17694.27, 1.639344262295082
52, 35481, 17740.5, 2
53, 43401, 17618.227722772277, 2.4634146341463414
54, 53088, 17696, 3
55, 64938, 17533.26, 3.7037037037037037
56, 79433, 17132.607843137255, 4.636363636363637
57, 97163, 17750.93269230769, 5.473684210526316
58, 118850, 17650.9900990099, 6.733333333333333
59, 145378, 17499.203703703704, 8.307692307692308
60, 177828, 17782.8, 10
61, 217520, 17797.090909090908, 12.222222222222221
62, 266073, 17738.2, 15
63, 325462, 17752.47272727273, 18.333333333333332
64, 398107, 17772.633928571428, 22.4
65, 486968, 17086.59649122807, 28.5
66, 595662, 17692.930693069306, 33.666666666666664
67, 728618, 17771.170731707316, 41
68, 891251, 17648.534653465347, 50.5
69, 1090184, 17726.56910569106, 61.5
70, 1333521, 17780.28, 75
71, 1631173, 17730.141304347828, 92
72, 1995262, 17053.521367521367, 117
73, 2440619, 17814.73722627737, 137
74, 2985383, 17664.988165680472, 169
75, 3651741, 17726.898058252427, 206
76, 4466836, 17796.159362549803, 251
77, 5463865, 17568.69774919614, 311
78, 6683439, 17822.504, 375
79, 8175230, 17810.958605664488, 459
80, 10000000, 17730.49645390071, 564
```

Looking at 250% improvement in performance.
Probably won't be consistent,  but,  I doubt there will ever be a performance *loss* given the simplicity of the change.

### CODE REVIEW

File: sha256.js [hunk 1]
```diff
@@ -53,37 +53,12 @@ Sha256.prototype.init = function () {
...
-
-function Gamma0256 (x) {
-  return (S(x, 7) ^ S(x, 18) ^ R(x, 3))
-}
-
-function Gamma1256 (x) {
-  return (S(x, 17) ^ S(x, 19) ^ R(x, 10))
-}
+function Ch (x, y, z) { return z ^ (x & (y ^ z)) }
+function Maj (x, y, z) { return (x & y) | (z & (x | y)) }
```
Comment (dcousens) [inline sha256.js]: Was also able to reduce these expressions a little bit

**Comment by dcousens** [comment]
Will do this for SHA256, hold up

**Comment by dcousens** [comment]
Performance improvement was basically 0% for SHA512,  but,  for consistency I kept the changes.

File: sha256.js [hunk 2]
```diff
@@ -53,36 +53,28 @@ Sha256.prototype.init = function () {
...
-}
-
 function Ch (x, y, z) {
-  return ((x & y) ^ ((~x) & z))
+  return z ^ (x & (y ^ z))
 }
 
 function Maj (x, y, z) {
-  return ((x & y) ^ (x & z) ^ (y & z))
+  return (x & y) | (z & (x | y))
```
Comment (dcousens) [inline sha256.js]: Was also able to reduce these expressions a little bit

**Comment by dominictarr** [comment]
Funny that V8 couldn't figure this out on it's own! these are named functions and have no free variables. it would be straightforward to have a preprocessor that made substitutions like this.

**Comment by dominictarr** [comment]
I'm seeing a much smaller difference. from about 12k bytes/second to 14k. this is with iojs@2.3.1 what version are you using?

**Comment by dcousens** [comment]
@dominictarr for these tests? `node v0.12.7`

**Comment by dominictarr** [comment]
okay so my result for your branch is about the same, but my master branch is 2x as fast as you. what commit are you testing against?

**Comment by dominictarr** [comment]
installing nvm so i can test with other versions...

**Comment by dcousens** [comment]
@dominictarr I was testing before and after.  Literally.  Stash, no stash.

**Comment by dcousens** [comment]
Any blockers to merge?

**Comment by dominictarr** [comment]
This is an improvement I am seeing 11% improvement, but I am just curious about why you are seeing much more divergent results than I am. @dcousens could you run the benchmark again and see if you get similar results.

the evidence I have gathered is still suggesting this is worthwhile merging, but there seems to be something we don't understand going on and that is really bugging me.

**Comment by dcousens** [comment]
``` bash
node -v
  v0.12.7

git clone https://github.com/crypto-browserify/sha.js
git clone https://github.com/dominictarr/crypto-bench
cd sha.js
npm install
npm link
cd ../crypto-bench
npm install
npm link sha.js
node bench-hash sha.js sha256 > before.log
cd ../sha.js
git checkout perf
cd ../crypto-bench
node bench-hash sha.js sha256 > after.log
```

before.log
``` 
run (N), input-size (bytes), ops (bytes/ms), time (ms)
run (N), input-size (bytes), ops (bytes/ms), time (ms)
1, 1, 69.04, 0.014484356894553883
3, 2, 166.7, 0.01199760047990402
5, 3, 259.23, 0.011572734637194769
7, 4, 349.92, 0.011431184270690443
8, 5, 427.65, 0.0116918040453642
9, 6, 528.6, 0.011350737797956867
10, 7, 616.56, 0.011353315168029064
11, 9, 793.17, 0.01134687393623057
12, 11, 965.69, 0.011390818999886091
13, 14, 1234.52, 0.011340440009072352
14, 17, 1454.01, 0.0116918040453642
15, 21, 1844.01, 0.011388224575788634
16, 25, 2192.75, 0.011401208528103978
17, 31, 2729.86, 0.011355893708834885
18, 38, 3337.54, 0.011385631333257429
19, 46, 4042.48, 0.01137915339098771
20, 56, 2653.28, 0.021105951878429716
21, 69, 3307.17, 0.020863759649488837
22, 84, 4026.12, 0.020863759649488837
23, 103, 4764.78, 0.021616947686986597
24, 126, 4147.92, 0.030376670716889428
25, 154, 5055.82, 0.03045994517209869
26, 188, 4700, 0.04
27, 230, 5750, 0.04
28, 282, 5704.86, 0.049431537320810674
29, 345, 5847.75, 0.058997050147492625
30, 422, 6178.08, 0.06830601092896176
31, 516, 5913.36, 0.08726003490401396
32, 631, 6278.45, 0.10050251256281408
33, 772, 6160.56, 0.12531328320802004
34, 944, 6523.04, 0.1447178002894356
35, 1155, 6329.4, 0.18248175182481752
36, 1413, 6386.76, 0.22123893805309736
37, 1728, 6445.44, 0.2680965147453083
38, 2113, 6486.91, 0.3257328990228013
39, 2585, 6617.6, 0.390625
40, 3162, 6608.58, 0.4784688995215311
41, 3868, 6575.6, 0.5882352941176471
42, 4732, 6530.16, 0.7246376811594203
43, 5788, 6656.2, 0.8695652173913043
44, 7079, 6725.05, 1.0526315789473684
45, 8660, 6754.8, 1.2820512820512822
46, 10593, 6673.59, 1.5873015873015872
47, 12957, 6737.64, 1.9230769230769231
48, 15849, 6747.59405940594, 2.3488372093023258
49, 19387, 6718.267326732674, 2.8857142857142857
50, 23714, 6676.757281553398, 3.5517241379310347
51, 29007, 6540.794117647059, 4.434782608695652
52, 35481, 6758.285714285715, 5.25
53, 43401, 6741.902912621359, 6.4375
54, 53088, 6700.42718446602, 7.923076923076923
55, 64938, 6738.8490566037735, 9.636363636363637
56, 79433, 6744.311320754717, 11.777777777777779
57, 97163, 6734.069306930693, 14.428571428571429
58, 118850, 6791.428571428572, 17.5
59, 145378, 6793.3644859813085, 21.4
60, 177828, 6586.222222222223, 27
61, 217520, 6692.923076923077, 32.5
62, 266073, 6764.567796610169, 39.333333333333336
63, 325462, 6733.696551724138, 48.333333333333336
64, 398107, 6747.576271186441, 59
65, 486968, 6716.8, 72.5
66, 595662, 6768.886363636364, 88
67, 728618, 6564.126126126126, 111
68, 891251, 6751.901515151515, 132
69, 1090184, 6771.329192546584, 161
70, 1333521, 6734.954545454545, 198
71, 1631173, 6768.352697095435, 241
72, 1995262, 6650.873333333333, 300
73, 2440619, 6779.497222222222, 360
74, 2985383, 6754.260180995475, 442
75, 3651741, 6737.529520295203, 542
76, 4466836, 6747.486404833837, 662
77, 5463865, 6795.851990049751, 804
78, 6683439, 6730.552870090634, 993
79, 8175230, 6728.584362139918, 1215
80, 10000000, 6765.899864682003, 1478
```

after.log
```
run (N), input-size (bytes), ops (bytes/ms), time (ms)
1, 1, 148.89, 0.0067163677883000875
3, 2, 365.46, 0.005472555135992995
5, 3, 543.63, 0.005518459246178467
7, 4, 748.64, 0.005343022013250694
8, 5, 894.4, 0.005590339892665474
9, 6, 1118.94, 0.005362217813287576
10, 7, 1307.04, 0.005355612682090831
11, 9, 1675.62, 0.005371146202599634
12, 11, 2053.59, 0.005356473297980609
13, 14, 2608.62, 0.005366822304513497
14, 17, 3164.38, 0.0053723004190394325
15, 21, 3910.2, 0.0053705692803437165
16, 25, 4632, 0.005397236614853195
17, 31, 5697.8, 0.00544069640914037
18, 38, 6458.86, 0.005883391186680002
19, 46, 8418.92, 0.005463883728554256
20, 56, 6208.16, 0.009020386072523904
21, 69, 7639.68, 0.009031791907514452
22, 84, 9262.68, 0.009068649678062937
23, 103, 11386.65, 0.009045680687471733
24, 126, 9948.96, 0.012664640324214792
25, 154, 12136.74, 0.01268874508311128
26, 188, 11595.84, 0.01621271076523995
27, 230, 13542.4, 0.016983695652173912
28, 282, 14119.74, 0.019972039145196723
29, 345, 14638.35, 0.023568230025925053
30, 422, 15487.4, 0.027247956403269755
31, 516, 15000.12, 0.03439972480220158
32, 631, 16595.3, 0.03802281368821293
33, 772, 15818.28, 0.04880429477794046
34, 944, 16822.08, 0.05611672278338945
35, 1155, 16481.85, 0.0700770847932726
36, 1413, 15952.77, 0.08857395925597875
37, 1728, 16778.88, 0.10298661174047374
38, 2113, 16925.13, 0.12484394506866417
39, 2585, 17319.5, 0.14925373134328357
40, 3162, 17359.38, 0.18214936247723132
41, 3868, 17406, 0.2222222222222222
42, 4732, 17082.52, 0.2770083102493075
43, 5788, 17479.76, 0.33112582781456956
44, 7079, 17626.71, 0.40160642570281124
45, 8660, 17320, 0.5
46, 10593, 17584.38, 0.6024096385542169
47, 12957, 17232.81, 0.7518796992481203
48, 15849, 17750.88, 0.8928571428571429
49, 19387, 17659.445544554455, 1.0978260869565217
50, 23714, 17785.5, 1.3333333333333333
51, 29007, 17694.27, 1.639344262295082
52, 35481, 17916.148514851484, 1.9803921568627452
53, 43401, 17794.41, 2.4390243902439024
54, 53088, 17696, 3
55, 64938, 17189.470588235294, 3.7777777777777777
56, 79433, 17737.466019417476, 4.478260869565218
57, 97163, 17750.93269230769, 5.473684210526316
58, 118850, 17827.5, 6.666666666666667
59, 145378, 17829.377358490565, 8.153846153846153
60, 177828, 17782.8, 10
61, 217520, 17797.090909090908, 12.222222222222221
62, 266073, 17738.2, 15
63, 325462, 17915.339449541283, 18.166666666666668
64, 398107, 17460.833333333332, 22.8
65, 486968, 17870.38532110092, 27.25
66, 595662, 17869.86, 33.333333333333336
67, 728618, 17916.83606557377, 40.666666666666664
68, 891251, 17825.02, 50
69, 1090184, 17871.868852459018, 61
70, 1333521, 17899.610738255033, 74.5
71, 1631173, 17539.494623655915, 93
72, 1995262, 17975.333333333332, 111
73, 2440619, 17814.73722627737, 137
74, 2985383, 17876.54491017964, 167
75, 3651741, 17900.691176470587, 204
76, 4466836, 17939.100401606425, 249
77, 5463865, 17739.821428571428, 308
78, 6683439, 18014.66037735849, 371
79, 8175230, 17810.958605664488, 459
80, 10000000, 17889.08765652952, 559
```

**Comment by dominictarr** [comment]
okay well, that is weird but I guess we can say it's 10-60 percent improvement.
merging

**Comment by dominictarr** [comment]
merged into 2.4.3

**Comment by dcousens** [comment]
@dominictarr that was weird. 
I viewed the commit history on my phone, and only showed https://github.com/crypto-browserify/sha.js/commit/2047818ca6309f627a2743c28b6b8f83c51e8899,  with https://github.com/crypto-browserify/sha.js/commit/ee9184bb71cb672167b5f43f580ccb0cf9c22402 right before it.

On merge, they all appeared?
Sorry, I can see now you merged it manually.

**Comment by dominictarr** [comment]
that was weird. I manually checked the files look right after your merge and published it as 2.4.4

**Pull request merged**